pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -659,7 +663,7 @@ func pickMatchingGatewayHosts(gatewayServerHosts map[model.Hostname]bool, virtua
 			if gwHostnameForMatching.Matches(model.Hostname(vsvcHost)) {
 				// assign the actual gateway host because calling code uses it as a key
 				// to locate TLS redirect servers
-				matchingHosts[vsvcHost] = gatewayHost
+				matchingHosts[vsvcHost] = append(matchingHosts[vsvcHost], gatewayHost)
```

Comment:
for line 663, is there any reason it is a Matches() not a SubSetOf() so that you don't need to call SubSetOf() again in line 264?  It will also take care of buildGatewayNetworkFiltersFromTLSRoutes().

Reply:
It needs to call Matches because the match can be in either direction.

All I did in this PR is add code to fix #12655, but @rshriram mentioned to me that he thinks this code could use a much bigger rework and cleanup. I'll defer to him, whether we want to merge this PR for the bug fix and cleanup in a later PR, or if he feels strongly we should cleanup right way.

@rshriram WDYT?

Reply:
Is there a scenario that we need to use the other direction for Gateway and VS host merging?  I had thought it is always Gateway to VS but perhaps I'm wrong.   Are you saying it is whichever the host field is tighter that would win?

Reply:
exactly

Reply:
btw, I think the api doc should explain this more clearly, but I don't want to fix the doc until the bug is fixed and it actually works. Otherwise it will just help highlight the bug :)

Reply:
ok, i'd vote to get this in, unless the bigger rework is happening from @rshriram really soon.